assets/src/loader.rs

```diff
@@ -0,0 +1,348 @@
+          N: Into<String>,
+          S: Store
+{
+    let name = name.into();
+
+    let store_id = storage.store_id();
+    let spec = AssetSpec::new(name.clone(), F::extension(), store_id);
+
+    A::retrieve(context, &spec)
+        .map(|a| Ok(a))
+        .unwrap_or_else(move || load_asset_inner(context, format, spec, storage))
+        .map_err(|e| AssetError::new(AssetSpec::new(name, F::extension(), store_id), e))
+}
+
+/// Loads an asset with a given context, format, specifier and storage right now.
+pub fn reload_asset<A, F, N, S>(context: &A::Context,
+                                format: &F,
+                                name: N,
+                                storage: &S)
+                                -> Result<A, AssetError<A::Error, F::Error, S::Error>>
+    where A: Asset,
+          F: Format<Data = A::Data>,
+          N: Into<String>,
+          S: Store
+{
+    let name = name.into();
+
+    let store_id = storage.store_id();
+    let spec = AssetSpec::new(name.clone(), F::extension(), store_id);
+
+    load_asset_inner(context, format, spec, storage)
+        .map_err(|e| AssetError::new(AssetSpec::new(name, F::extension(), store_id), e))
+}
+
+fn load_asset_inner<A, F, S>(context: &A::Context,
+                             format: &F,
+                             spec: AssetSpec,
+                             store: &S)
+                             -> Result<A, LoadError<A::Error, F::Error, S::Error>>
+    where A: Asset,
+          F: Format<Data = A::Data>,
+          S: Store
+{
+    let bytes = store
+        .load(A::category(), &spec.name, spec.ext)
+        .map_err(LoadError::StorageError)?;
+    let data = format.parse(bytes).map_err(LoadError::FormatError)?;
+    let a = Asset::from_data(data, context)
+        .map_err(LoadError::AssetError)?;
+
+    A::cache(context, spec, &a);
+
+    Ok(a)
+}
+
+/// Like `load_asset`, but loads the asset on a worker thread and returns
+/// an `AssetFuture` immediately.
+pub fn load_asset_future<A, F, N, S>(context: Arc<A::Context>,
+                                     format: F,
+                                     name: N,
+                                     storage: Arc<S>,
+                                     thread_pool: &ThreadPool)
+                                     -> AssetFuture<A, AssetError<A::Error, F::Error, S::Error>>
+    where A: Asset + Send + 'static,
+          A::Context: Send + Sync + 'static,
+          A::Error: Send + 'static,
+          F: Format<Data = A::Data> + Send + 'static,
+          F::Error: Send + 'static,
+          N: Into<String>,
+          S: Store + Send + Sync + 'static
+{
+    let name = name.into();
+
+    let closure = move || load_asset::<A, F, _, S>(&*context, &format, name, &*storage);
+
+    AssetFuture::spawn(thread_pool, closure)
+}
+
+/// Like `load_asset`, but loads the asset on a worker thread and returns
+/// an `AssetFuture` immediately.
+pub fn reload_asset_future<A, F, N, S>
```

Comment:
Should these `reload` methods specify the difference between what they do and just the regular `load` methods?